src/v/kafka/client/client.cc

```diff
@@ -174,6 +178,65 @@ ss::future<produce_response::partition> client::produce_record_batch(
       });
 }
 
+ss::future<produce_response> client::produce_records(
+  model::topic topic, std::vector<record_essence> records) {
+    absl::node_hash_map<model::partition_id, storage::record_batch_builder>
+      partition_builders;
+
+    // Assign records to batches per topic_partition
+    for (auto& record : records) {
+        auto p_id = record.partition_id;
+        if (!p_id) {
+            p_id = co_await gated_retry_with_mitigation([&, this]() {
+                       return _topic_cache.partition_for(topic, record);
+                   }).handle_exception_type([](const topic_error&) {
+                // Assume auto topic creation is on and assign to first
+                // partition
+                return model::partition_id{0};
+            });
+        }
+        auto it = partition_builders.find(*p_id);
+        if (it == partition_builders.end()) {
+            it = partition_builders
+                   .emplace(
+                     *p_id,
+                     storage::record_batch_builder(
+                       raft::data_batch_type, model::offset(0)))
+                   .first;
+        }
+        it->second.add_raw_kw(
+          std::move(record.key).value_or(iobuf{}),
```

Comment:
should the move be around the entire `optional.value_or` expression?

Reply:
There's an rvalue-reference overload for `value_or`

Reply:
very neat